### PR TITLE
feat: update the Dockerfile with emulator dependencies

### PR DESCRIPTION
**Pull request by renan061**

### CODE REVIEW

File: build/Dockerfile [hunk 1]
```diff
@@ -4,170 +4,301 @@
...
 
-# STAGE: rollups-node-snapshot
+# =============================================================================
+# TARGET: rollups-node-snapshot
 #
-# This stage copies the image from the builder.
-# We use the emulator as base image so we can easily create a container with a volume shared with
-# the rollups-node container.
-FROM emulator-base as rollups-node-snapshot
+# TODO: this should be deleted.
```
Comment (marcelstanley) [inline build/Dockerfile:223]: Please remove the TODO. What's the reason behind it? We should either cover this in a separate ticket.

File: build/Dockerfile [hunk 2]
```diff
@@ -4,170 +4,301 @@
...
+ARG MACHINE_TOOLS_VERSION
+ARG MACHINE_IMAGE_KERNEL_VERSION
+ARG MACHINE_KERNEL_VERSION
 ARG DEBIAN_FRONTEND=noninteractive
+
+# Install ca-certificates and curl (setup).
+RUN <<EOF
+    set -e
+    apt-get update
+    apt-get install -y --no-install-recommends ca-certificates curl
```
Comment (marcelstanley) [inline build/Dockerfile:54]:
```suggestion
    apt-get install -y --no-install-recommends \
        ca-certificates \
        curl
```

File: build/docker-bake.platforms.hcl [hunk 3]
```diff
@@ -4,6 +4,6 @@
 target "docker-platforms" {
     platforms = [
         "linux/amd64",
-        "linux/arm64"
+        # "linux/arm64"
```
Comment (marcelstanley) [inline build/docker-bake.platforms.hcl:7]: Perhaps we should at least create a ticket for this or document it elsewhere.

```suggestion
        #unsupported due to libcmt "linux/arm64"
```

File: build/Dockerfile [hunk 4]
```diff
@@ -4,170 +4,301 @@
...
+
+# =============================================================================
+# STAGE: emulator-devel
+#
+# - Install libarchive13 (setup -- required by xgenext2fs).
+# - Install libcmt.
+# - Install xgenext2fs.
+#
+# IMPORTANT: Only works with the linux/amd64 platform!
+# =============================================================================
```
Comment (marcelstanley) [inline build/Dockerfile:116]: Let's keep the current convention to differentiate between targets and stages:

```suggestion
# STAGE: emulator-devel
#
# - Install libarchive13 (setup -- required by xgenext2fs).
# - Install libcmt.
# - Install xgenext2fs.
#
# IMPORTANT: Only works with the linux/amd64 platform!
```

File: build/Dockerfile [hunk 5]
```diff
@@ -4,170 +4,301 @@
...
+EOF
+
+# =============================================================================
+# STAGE: emulator-devel
+#
+# - Install libarchive13 (setup -- required by xgenext2fs).
+# - Install libcmt.
+# - Install xgenext2fs.
+#
+# IMPORTANT: Only works with the linux/amd64 platform!
```
Comment (marcelstanley) [inline build/Dockerfile:115]: Perhaps we should tell which libraries add this constraint.

File: build/Dockerfile [hunk 6]
```diff
@@ -4,170 +4,301 @@
...
+        docker-ce \
+        docker-ce-cli \
+        containerd.io \
+        docker-buildx-plugin \
+        docker-compose-plugin
+    # Cartesi Machine Emulator
+    SDK_URL=https://github.com/cartesi/machine-emulator-sdk
+EOF
+
+# =============================================================================
```
Comment (marcelstanley) [inline build/Dockerfile:203]:
```suggestion
```

File: build/Dockerfile [hunk 7]
```diff
@@ -4,170 +4,301 @@
...
-# Download rootfs and linux.
-# Add these files to the directories the cartesi-machine expects.
-WORKDIR /usr/share/cartesi-machine/images/
-ARG TOOLS_VERSION
-ARG LINUX_VERSION
-ARG LINUX_KERNEL_VERSION
-ADD https://github.com/cartesi/machine-emulator-tools/releases/download/v${TOOLS_VERSION}/rootfs-tools-v${TOOLS_VERSION}.ext2 rootfs.ext2
-ADD https://github.com/cartesi/image-kernel/releases/download/v${LINUX_VERSION}/linux-${LINUX_KERNEL_VERSION}.bin linux.bin
+# This stage builds an echo snapshot.
+# =============================================================================
```
Comment (marcelstanley) [inline build/Dockerfile:207]:
```suggestion
```

File: build/Dockerfile [hunk 8]
```diff
@@ -4,170 +4,301 @@
...
 # Set dummy command.
 CMD /bin/bash
 
-####################################################################################################
-# TARGET: rollups-node-devnet
-#
-# This target contains the Ethereum node that rollups node uses for testing.
-# This target requires the machine-snapshot built in the snapshot-builder stage.
-
+# =============================================================================
```
Comment (marcelstanley) [inline build/Dockerfile:238]:
```suggestion
```

File: build/Dockerfile [hunk 9]
```diff
@@ -4,170 +4,301 @@
...
 # STAGE: devnet-base
 #
 # This stage installs Foundry.
+# =============================================================================
```
Comment (marcelstanley) [inline build/Dockerfile:242]:
```suggestion
```

File: build/Dockerfile [hunk 10]
```diff
@@ -217,47 +354,52 @@ HEALTHCHECK --interval=1s --timeout=1s --retries=5 \
 # Start Anvil.
 CMD anvil --block-time 1 --load-state $ANVIL_STATE_PATH
 
-####################################################################################################
-# TARGET: rollups-node
-#
-# This target is the final Docker image that will be used in the production environment.
-# It installs all the binaries necessary to run the node in /usr/bin.
-
+# =============================================================================
```
Comment (marcelstanley) [inline build/Dockerfile:357]:
```suggestion
```

File: build/Dockerfile [hunk 11]
```diff
@@ -217,47 +354,52 @@ HEALTHCHECK --interval=1s --timeout=1s --retries=5 \
...
 # STAGE: rust-chef
 #
 # This stage installs the cargo-chef tool.
+# =============================================================================
```
Comment (marcelstanley) [inline build/Dockerfile:361]:
```suggestion
```

File: build/Dockerfile [hunk 12]
```diff
@@ -217,47 +354,52 @@ HEALTHCHECK --interval=1s --timeout=1s --retries=5 \
...
 RUN <<EOF
-set -e
-rustup component add rustfmt
-cargo install cargo-chef
+    set -e
+    rustup component add rustfmt
+    cargo install cargo-chef
 EOF
 
+# =============================================================================
```
Comment (marcelstanley) [inline build/Dockerfile:372]:
```suggestion
```

File: build/Dockerfile [hunk 13]
```diff
@@ -217,47 +354,52 @@ HEALTHCHECK --interval=1s --timeout=1s --retries=5 \
...
 # STAGE: rust-prepare
 #
 # This stage prepares the recipe with just the external dependencies.
+# =============================================================================
+
 FROM rust-chef as rust-prepare
 COPY ./cmd/authority-claimer/ .
 RUN cargo chef prepare --recipe-path recipe.json
 
+# =============================================================================
```
Comment (marcelstanley) [inline build/Dockerfile:382]:
```suggestion
```

File: build/Dockerfile [hunk 14]
```diff
@@ -268,10 +410,13 @@ RUN cargo chef cook --release --recipe-path recipe.json
 COPY ./cmd/authority-claimer/ .
 RUN cargo build --release
 
+# =============================================================================
```
Comment (marcelstanley) [inline build/Dockerfile:413]:
```suggestion
```

File: build/Dockerfile [hunk 15]
```diff
@@ -268,10 +410,13 @@ RUN cargo chef cook --release --recipe-path recipe.json
...
 # STAGE: go-builder
 #
 # This stage builds the Go binaries.
 # First it downloads the external dependencies and then it builds the binaries.
+# =============================================================================
+
```
Comment (marcelstanley) [inline build/Dockerfile:419]:
```suggestion
```

File: build/Dockerfile [hunk 16]
```diff
@@ -4,170 +4,301 @@
...
+# - Install docker.
+# =============================================================================
+
+FROM emulator-devel as rollups-node-ci
+
+# Install git and build-essential (setup).
+RUN <<EOF
+    set -e
+    apt-get update
+    apt-get install -y --no-install-recommends git-all build-essential
```
Comment (marcelstanley) [inline build/Dockerfile:177]:
```suggestion
    apt-get install -y --no-install-recommends \
        git-all \
        build-essential
```

File: build/Dockerfile [hunk 17]
```diff
@@ -4,170 +4,301 @@
...
 RUN <<EOF
-set -e
-apt-get update
-apt-get install -y --no-install-recommends \
-    ca-certificates \
-    curl \
-    git
+    set -e
+    apt-get update
+    apt-get install -y --no-install-recommends ca-certificates curl git
```
Comment (marcelstanley) [inline build/Dockerfile:251]:
```suggestion
    apt-get install -y --no-install-recommends \
        ca-certificates \
        curl \
        git
```

File: build/Dockerfile [hunk 18]
```diff
@@ -4,170 +4,301 @@
...
+RUN <<EOF
+    set -e
+    URL=https://github.com/foundry-rs/foundry/releases/download
+    VERSION=nightly-${FOUNDRY_NIGHTLY_VERSION}
+    ARCH=$(dpkg --print-architecture)
+    ARTIFACT=foundry_nightly_linux_${ARCH}.tar.gz
+    curl -sSL ${URL}/${VERSION}/${ARTIFACT} | tar -zx -C /usr/local/bin
+EOF
 
+# =============================================================================
```
Comment (marcelstanley) [inline build/Dockerfile:265]:
```suggestion
```

File: build/Dockerfile [hunk 19]
```diff
@@ -4,170 +4,301 @@
...
 # STAGE: devnet-builder
 #
 # This stage builds the devnet state that will be loaded in Anvil.
+# =============================================================================
+
```
Comment (marcelstanley) [inline build/Dockerfile:270]:
```suggestion
```

File: build/Dockerfile [hunk 20]
```diff
@@ -217,47 +354,52 @@ HEALTHCHECK --interval=1s --timeout=1s --retries=5 \
...
+    rustup component add rustfmt
+    cargo install cargo-chef
 EOF
 
+# =============================================================================
 # STAGE: rust-prepare
 #
 # This stage prepares the recipe with just the external dependencies.
+# =============================================================================
+
```
Comment (marcelstanley) [inline build/Dockerfile:377]:
```suggestion
```

File: build/Dockerfile [hunk 21]
```diff
@@ -217,47 +354,52 @@ HEALTHCHECK --interval=1s --timeout=1s --retries=5 \
...
+# =============================================================================
 # STAGE: rust-builder
 #
 # This stage builds the Rust binaries.
-# First it builds the external dependencies and then it builds the node binaries.
+#
+# First it builds the external dependencies and then it builds the node
+# binaries.
+# =============================================================================
+
```
Comment (marcelstanley) [inline build/Dockerfile:390]:
```suggestion
```

File: build/Dockerfile [hunk 22]
```diff
@@ -286,33 +431,40 @@ RUN go mod download
...
+# =============================================================================
+# TARGET: rollups-node
+#
+# This target is the final Docker image that will be used in the production
+# environment. It installs all the binaries necessary to run the node in
+# /usr/bin.
 #
-# This stage copies the binaries from the previous stages.
-FROM emulator-base as rollups-node
+# The stage copies the binaries from previous stages.
```
Comment (marcelstanley) [inline build/Dockerfile:441]:
```suggestion
# This stage copies the binaries from previous stages.
```

File: pkg/addresses/addresses.go [hunk 23]
```diff
@@ -35,7 +35,7 @@ type Book struct {
 // Get the addresses for the test environment.
 func GetTestBook() *Book {
 	return &Book{
-		Application:         common.HexToAddress("0xb72c832dDeA10326143831F1E5F1646920C9c990"),
+		Application:         common.HexToAddress("0x2E663fe9aE92275242406A185AA4fC8174339D3E"),
```
Comment (marcelstanley) [inline pkg/addresses/addresses.go:38]: Please move all updates to the application address to a separate commit. If possible, that commit should also contain the update of the emulator SDK version.

**Review by marcelstanley — COMMENTED** [review]
I'm assuming, it would be preferrable to have a single commit for most of the changes related to the emulator and the CI. In that case, it would be helpful it the commit contained a more thorough explanation about their purpose as the commit message does not convey much about it.

Also, in order to make the changes clearer in the diff, I'd suggest either creating a separate commit for the formatting changes or reverting them altogether. This would help a lot with reviewing `Dockerfile` and `build/docker-bake.hcl` (it's a bit hard to identify the upgrades being made to the versions)

File: build/Dockerfile (hunk 1 above)
Comment (renan061) [reply build/Dockerfile:223]: Serves as a reminder (and temporary documentation), that we should not create snapshots in the CI anymore. I'm going to update it to add this information.

File: build/Dockerfile (hunk 2 above)
Comment (renan061) [reply build/Dockerfile:54]: I think we should avoid breaking lines unless they are longer than the 80 characters limit.
It makes the Dockerfile more readable.
The fact that we are installing ca-certificates and curl is already explained in a commentary.

File: build/docker-bake.platforms.hcl (hunk 3 above)
Comment (renan061) [reply build/docker-bake.platforms.hcl:7]: Changed to:
```
# TODO: libarchive13 (required by xgenext2fs in the emulator-devel
# stage) is not available for arm64. We are temporarily disabling this
# platform now, but will come back to it before merging next/2.0 into
# main.
# "linux/arm64"
```

File: build/Dockerfile (hunk 5 above)
Comment (renan061) [reply build/Dockerfile:115]: Changed to:
```
# IMPORTANT: Only works with the linux/amd64 platform!
#                         (libarchive13, required by xgenext2fs, is amd64 exclusive.)
```

File: build/Dockerfile (hunk 17 above)
Comment (renan061) [reply build/Dockerfile:251]: Same as before.

File: build/Dockerfile (hunk 4 above)
Comment (renan061) [reply build/Dockerfile]: Now thinking about it, I think we should only have stages in the Dockerfile.
Targets should be left to the `docker-bake` files.

**Comment by renan061** [comment]
> I'm assuming it would be preferable to have a single commit for most of the changes related to the emulator and the CI. In that case, it would be helpful it the commit contained a more thorough explanation about their purpose as the commit message does not convey much about it.
> 
> Also, in order to make the changes clearer in the diff, I'd suggest either creating a separate commit for the formatting changes or reverting them altogether. This would help a lot with reviewing `Dockerfile` and `build/docker-bake.hcl` (it's a bit hard to identify the upgrades being made to the versions)

I've update the commit message with more information.

Separating the refactor in another commit would further delay this commit.
I think we can get together in a call for me to explain stuff.

File: build/docker-bake.platforms.hcl (hunk 3 above)
Comment (torives) [reply build/docker-bake.platforms.hcl:11]: Danilo [confirmed](https://github.com/cartesi/rollups-node/issues/439#issuecomment-2150398192) there's an arm64 release for xgenext2fs, so we should just adapt the PR accordingly and close #439.

**Review by torives — COMMENTED** [review]
Let's just reinstate the linux/arm64 platform now that [we know](https://github.com/cartesi/rollups-node/issues/439#issuecomment-2150398192) that xgenext2fs has a build for it. Other than that, it looks good to me 👌🏽

File: build/Dockerfile (hunk 2 above)
Comment (marcelstanley) [reply build/Dockerfile:54]: I'd rather have each item in their own lines. IMO, it makes it easier to read and `diff` future additions

**Comment by marcelstanley** [comment]
> Let's just reinstate the linux/arm64 platform now that [we know](https://github.com/cartesi/rollups-node/issues/439#issuecomment-2150398192) that xgenext2fs has a build for it. Other than that, it looks good to me 👌🏽

Please close #439 when this is done, @renan061

**Comment by renan061** [comment]
Fixed `platform/arm64`. Now working!

Regarding the documentation for targets, the same text is still there.
What changes is that we are not identifying which stages are targets in the Dockerfile anymore.

Because `docker-bake.hcl` is very explicit about which stages are targets,
I don't think we need to document it elsewhere.

**Review by torives — APPROVED** [review]

**Review by marcelstanley — APPROVED** [review]

**Pull request merged**